src/server/replica.cc

```diff
@@ -830,26 +830,30 @@ void DflyShardReplica::StableSyncDflyReadFb(Context* cntx) {
       break;
 
     last_io_time_ = Proactor()->GetMonotonicTimeNs();
-
+    size_t records = 1;
     if (tx_data->opcode == journal::Op::PING) {
       force_ping_ = true;
-      journal_rec_executed_.fetch_add(1, std::memory_order_relaxed);
+      const auto expect = journal_rec_executed_.load();
+      DCHECK(tx_data->lsn == expect)
+          << "tx_data->lsn=" << tx_data->lsn << " journal_rec_executed=" << expect;
     } else if (tx_data->opcode == journal::Op::EXEC) {
       if (use_multi_shard_exe_sync_) {
+        records = tx_data->journal_rec_count;
         InsertTxDataToShardResource(std::move(*tx_data));
       } else {
         // On no shard sync mode we execute multi commands once they are recieved, therefor when
         // receiving exec opcode, we only increase the journal counting.
         DCHECK_EQ(tx_data->commands.size(), 0u);
-        journal_rec_executed_.fetch_add(1, std::memory_order_relaxed);
       }
     } else {
+      records = tx_data->journal_rec_count;
       if (use_multi_shard_exe_sync_) {
         InsertTxDataToShardResource(std::move(*tx_data));
       } else {
         ExecuteTxWithNoShardSync(std::move(*tx_data), cntx);
       }
     }
+    journal_rec_executed_.fetch_add(records);
```

Comment:
I wrote in the issue that we need to compare with journal_rec_executed_, but actually now after reviewing the code I understand that we need a different counter to compare to which will be incremented inside NextTxData when reading another entry from socket. 
The reason for this is that under use_multi_shard_exe_sync_ we accumulate the multi transaction data and do not execute it therefor you might get different value in journal_rec_executed_ when comparing when getting a ping if it is in the middle of multi transaction

Reply:
> The reason for this is that under use_multi_shard_exe_sync_ we accumulate the multi transaction data and do not execute it therefor you might get different value in journal_rec_executed_

Exactly. That's why I moved the `fetch_add` from execute and replaced it here, so this doesn't happen. I am not 100% sure of this is correct but for now I am looking at why one of the replication tests triggers the DCHECK (and it's not because of multi)

Reply:
but journal_rec_executed_ was incremented after executing the command on purpose. replica sends master this value so that we will know the replica lag and that we can also do replica takeover only after lag is 0 meaning we will have no records that are not executed

Reply:
yes I understand this and that's why I said it's probably wrong. Ended up using a separate counter in `NextTxData` but we still have some issues :)